src/core/thread/mlr_manager.cpp

```diff
@@ -0,0 +1,406 @@
+    {
+        mTimer.Start(kTimerInterval);
+    }
+}
+
+void MlrManager::SendMulticastListenerRegistration(void)
+{
+    otError          error   = OT_ERROR_NONE;
+    Mle::MleRouter & mle     = Get<Mle::MleRouter>();
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+    IPv6AddressesTlv addressesTlv;
+    uint8_t          num;
+
+    VerifyOrExit(!mMlrPending, error = OT_ERROR_BUSY);
+    VerifyOrExit(mle.IsAttached(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(Get<BackboneRouter::Leader>().HasPrimary(), error = OT_ERROR_INVALID_STATE);
+
+    num = CountNetifMulticastAddresses(kMulticastAddressFilterNotRegistered);
+    VerifyOrExit(num > 0, error = OT_ERROR_NOT_FOUND);
+    if (num > kIPv6AddressesNumMax)
+    {
+        num = kIPv6AddressesNumMax;
+    }
+
+    VerifyOrExit((message = Get<Coap::Coap>().NewMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST);
+    SuccessOrExit(message->SetToken(Coap::Message::kDefaultTokenLength));
+    SuccessOrExit(message->AppendUriPathOptions(OT_URI_PATH_MLR));
+    SuccessOrExit(message->SetPayloadMarker());
+
+    addressesTlv.Init();
+    addressesTlv.SetLength(sizeof(Ip6::Address) * num);
+    SuccessOrExit(error = message->Append(&addressesTlv, sizeof(addressesTlv)));
+
+    for (Ip6::NetifMulticastAddress *addr = Get<ThreadNetif>().GetMulticastAddresses(); addr != nullptr;
+         addr                             = addr->GetNext())
+    {
+        if (Get<ThreadNetif>().IsMulticastAddressExternal(*addr) &&
+            addr->GetAddress().IsMulticastLargerThanRealmLocal() &&
+            MatchesFilter(*addr, kMulticastAddressFilterNotRegistered))
+        {
+            SuccessOrExit(error = message->Append(&addr->GetAddress(), sizeof(Ip6::Address)));
+            addr->SetMlrState(kMlrStateRegistering);
+
+            if (--num == 0)
+            {
+                break;
+            }
+        }
+    }
+
+    if (!mle.IsFullThreadDevice() && mle.GetParent().IsThreadVersion1p1())
+    {
+        uint8_t pbbrServiceId;
+
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetServiceId(pbbrServiceId));
+        SuccessOrExit(error = mle.GetServiceAloc(pbbrServiceId, messageInfo.GetPeerAddr()));
+    }
+    else
+    {
+        messageInfo.GetPeerAddr().SetToRoutingLocator(mle.GetMeshLocalPrefix(),
+                                                      Get<BackboneRouter::Leader>().GetServer16());
+    }
+
+    messageInfo.SetPeerPort(kCoapUdpPort);
+    messageInfo.SetSockAddr(mle.GetMeshLocal16());
+
+    SuccessOrExit(error = Get<Coap::Coap>().SendMessage(
+                      *message, messageInfo, &MlrManager::HandleMulticastListenerRegistrationResponse, this));
+
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        mMlrPending = true;
```

Comment:
I think it's better to put it after `exit` like this because it makes sure that `mMlrPending` is always correctly set to `true`. 
Currently it's fine for this function to put `mMlrPending = true;` before `exit:`. But if we add any statement like `VerifyOrExit(..., OT_NOOP)` or `ExitNow()` in this function, we are going to incorrectly skip setting `mMlrPending = true;`